### DISCUSSION TITLE
How to specify the uri of websocket

### DISCUSSION THREAD

**Discussion by coolbreeze1093** [resiprocate/resiprocate#316] [Q&A] [answered]
An address like this wss//:127.0.0.1:443

**Comment by coolbreeze1093** [comment]
WARNING | 20230614-170529.822 |  | RESIP | 21036 | Security.cxx:3122 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20230614-170529.823 |  | RESIP | 21036 | Security.cxx:3122 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
INFO | 20230614-170529.844 |  | RESIP:DNS | 21036 | AresDns.cxx:397 | DNS initialization: found  1 name servers
INFO | 20230614-170529.844 |  | RESIP:DNS | 21036 | AresDns.cxx:408 |  name server: 114.114.114.114
INFO | 20230614-170529.848 |  | RESIP:DNS | 21036 | DnsUtil.cxx:170 | local hostname does not contain a domain part sky
INFO | 20230614-170529.849 |  | RESIP | 21036 | Security.cxx:334 | Successfully loaded ./.sipCerts/domain_cert_client.pem
INFO | 20230614-170529.850 |  | RESIP | 21036 | Security.cxx:334 | Successfully loaded ./.sipCerts/domain_key_client.pemINFO | 20230614-170529.850 |  | RESIP | 21036 | Security.cxx:339 | Files loaded by prefix: 2
INFO | 20230614-170529.850 | basicClient.exe | RESIP:TRANSPORT | 21036 | Connection.cxx:45 | Connection::Connection: new connection created to who: [ V4 0.0.0.0:0 UNKNOWN_TRANSPORT ], is server = 0
INFO | 20230614-170529.851 | basicClient.exe | RESIP | 21036 | Security.cxx:440 | Security::updateDomainCtx: Successfully loaded domain cert, domain=client, filename=./.sipCerts/domain_cert_client.pem
INFO | 20230614-170529.851 | basicClient.exe | RESIP | 21036 | Security.cxx:466 | Security::updateDomainCtx: Successfully loaded domain private key, domain=client, filename=./.sipCerts/domain_key_client.pem
WARNING | 20230614-170529.851 | basicClient.exe | RESIP | 21036 | Security.cxx:3122 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
INFO | 20230614-170529.852 | basicClient.exe | RESIP:TRANSPORT | 21036 | WssTransport.cxx:47 | Creating WSS transport for domain client interface= port=5160
INFO | 20230614-170529.855 | basicClient.exe | RESIP:TRANSPORT | 21036 | TransportSelector.cxx:293 | TransportSelector::addTransport:  added transport for tuple=[ V4 0.0.0.0:5160 WSS targetDomain=client transportKey=1 ], key=1
INFO | 20230614-170529.855 | basicClient.exe | RESIP:TRANSPORT | 21036 | Connection.cxx:45 | Connection::Connection: new connection created to who: [ V4 0.0.0.0:0 UNKNOWN_TRANSPORT ], is server = 0
INFO | 20230614-170529.856 | basicClient.exe | RESIP | 21036 | Security.cxx:440 | Security::updateDomainCtx: Successfully loaded domain cert, domain=client, filename=./.sipCerts/domain_cert_client.pem
INFO | 20230614-170529.856 | basicClient.exe | RESIP | 21036 | Security.cxx:466 | Security::updateDomainCtx: Successfully loaded domain private key, domain=client, filename=./.sipCerts/domain_key_client.pem
WARNING | 20230614-170529.856 | basicClient.exe | RESIP | 21036 | Security.cxx:3122 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
INFO | 20230614-170529.857 | basicClient.exe | RESIP:TRANSPORT | 21036 | TlsTransport.cxx:44 | Creating TLS transport for domain client interface= port=5161
INFO | 20230614-170529.860 | basicClient.exe | RESIP:TRANSPORT | 21036 | TransportSelector.cxx:293 | TransportSelector::addTransport:  added transport for tuple=[ V4 0.0.0.0:5161 TLS targetDomain=client transportKey=2 ], key=2
INFO | 20230614-170529.862 | basicClient.exe | RESIP:TEST | 21036 | basicClientUserAgent.cxx:336 | register for sip:1006@27.22.247.231:2443
INFO | 20230614-170529.863 | basicClient.exe | RESIP:TEST | 21036 | basicClient.cxx:69 | E:\Github\resiprocate\x64\SSL-Debug-MT\basicClient.exe starting
INFO | 20230614-170529.865 | basicClient.exe | RESIP:TRANSPORT | 6820 | Connection.cxx:45 | Connection::Connection: new connection created to who: [ V4 27.22.247.231:2443 TLS targetDomain=27.22.247.231 flowKey=836 transportKey=2 ], is server = 0
INFO | 20230614-170529.865 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:73 | Creating TLS connection for domain client [ V4 27.22.247.231:2443 TLS targetDomain=27.22.247.231 transportKey=2 ] on 836
INFO | 20230614-170529.894 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:211 | TLS handshake starting (client mode)
INFO | 20230614-170529.895 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:221 | TLS connected
ERR | 20230614-170529.926 | basicClient.exe | RESIP | 6820 | Security.cxx:130 | Error when verifying peer's chain of certificates: self-signed certificate, depth=0 /C=CN/ST=Liaoning/L=Dalian/O=AVAYA/OU=APS/CN=lab

WARNING | 20230614-170529.926 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:291 | SSL cipher or certificate failure SSL_ERROR_SSL
ERR | 20230614-170529.926 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:321 | Server did not present any certificiate to us, certificate invalid or protocol did not reach certificate exchange
ERR | 20230614-170529.926 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:329 | TLS handshake failed
ERR | 20230614-170529.926 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:47 | error:0A000086:SSL routines::certificate verify failed
ERR | 20230614-170529.926 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:51 | Got TLS SSL_do_handshake error=1 ret=-1
ERR | 20230614-170529.926 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:164 | Unexpected error in SSL_shutdown
ERR | 20230614-170529.926 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:47 | error:0A000197:SSL routines::shutdown while in init
ERR | 20230614-170529.926 | basicClient.exe | RESIP:TRANSPORT | 6820 | TlsConnection.cxx:51 | Got TLS SSL_shutdown error=1 ret=-1
INFO | 20230614-170529.927 | basicClient.exe | RESIP:TRANSACTION | 12628 | TuSelector.cxx:102 | Sending ConnectionTerminated [ V4 27.22.247.231:2443 TLS targetDomain=27.22.247.231 flowKey=836 transportKey=2 ] to TUs
INFO | 20230614-170529.927 | basicClient.exe | RESIP:TEST | 21036 | basicClientUserAgent.cxx:444 | BasicClientUserAgent received connection terminated message for: [ V4 27.22.247.231:2443 TLS targetDomain=27.22.247.231 flowKey=836 transportKey=2 ]
INFO | 20230614-170529.927 | basicClient.exe | RESIP:TRANSACTION | 12628 | TransactionState.cxx:2380 | Try sending request to a different dns result
INFO | 20230614-170529.927 | basicClient.exe | RESIP:TRANSACTION | 12628 | TransactionState.cxx:2402 | No DNS results remain.
INFO | 20230614-170529.928 | basicClient.exe | RESIP:TRANSACTION | 12628 | TransactionState.cxx:2206 | Ran out of dns entries for 27.22.247.231. Send 503
INFO | 20230614-170529.928 | basicClient.exe | RESIP:TEST | 21036 | basicClientUserAgent.cxx:595 | onRequestRetry(ClientRegistrationHandle): msg=SipResp: 503 tid=6759874d44118c0d cseq=1 REGISTER / 1 from(wire), retryTime=35

The tls port of the server is 27443, the port of sip is 2443, I can't find a place to set the tls port

**Comment by sgodin** [comment]
I don't really understand the question or the setup, can you explain better?  What is meant by tls port vs sip port?  

As an off the cuff suggestion, I would recommend you look at the UserProfile setting setOutboundProxy, this controls where outbound SIP messages are routed to, and avoids looking up the destination from the Request-Uri.

https://github.com/resiprocate/resiprocate/wiki/DUM-Configuration-and-Profiles

**Reply by coolbreeze1093** [reply]
Thanks a lot for your answer! I am still relatively new to resip, and now I am debugging and learning basicClient. The problem I encountered is that my server is websocket+freeswitch, the websocket address is wss://127.0.0.1:27443, the sip address is 1005@127.0.0.1:2443, only the sip address can be set in the program, and the websocket cannot be set address, causing tls to only access port 2443. I don't know if I made it clear

**Reply by coolbreeze1093** [reply]
Suppose I am now modifying the resip, modifying the relevant parts of the network, and introducing the websocketpp library, do you think this is feasible?

**Comment by sgodin** [comment]
My suggestion is still valid to use the outbound proxy setting.  Try setting it to your websocket URL wss://127.0.0.1:27443

I don't see why you would bring websocketapp library into resip when it already has a websocket transport built in.  If anything we should improve / fixup the internal websocket implementation if something is not working.  I've done a lot work myself with server side socket connections on resip/repro (and it is working great), but I haven't played much with making an outbound/client side connection.  I believe others have done this though.

Scott

**Reply by coolbreeze1093** [reply]
Thank you very much for your suggestion, I will study the settings related to the outbound proxy

**Reply by coolbreeze1093** [reply]
As you said, I did outbound proxy settings like this
![image](https://github.com/resiprocate/resiprocate/assets/101555502/2d9c26d5-2246-49bb-b526-f84c766d99c9)

still get an error

INFO | 20230616-105733.402 | basicClient.exe | RESIP:TRANSACTION | 24692 | TransactionState.cxx:2206 | Ran out of dns entries for //27.22.247.231. Send 503
INFO | 20230616-105733.402 | basicClient.exe | RESIP:TEST | 16232 | basicClientUserAgent.cxx:596 | onRequestRetry(ClientRegistrationHandle): msg=SipResp: 503 tid=bc0bf3501110311e cseq=4 REGISTER / 4 from(wire), retryTime=408

**Reply by coolbreeze1093** [reply, marked as answer]
correct solution：
mOutboundEnabled = true;

//addTransport(UDP, mUdpPort);
   //addTransport(TCP, mTcpPort);
#if defined(USE_SSL)
   addTransport(TLS, mTlsPort);
#endif
#if defined(USE_DTLS)
   addTransport(DTLS, mDtlsPort);
#endif
   addTransport(WSS,5162);

Tuple t1(Data("127.0.0.1"), 27443, WSS);
   t1.mFlowKey = 1;
   mProfile->setClientOutboundFlowTuple(t1);

@sgodin Thank you!
